### PR TITLE
Add an app API method to fetch the system ColorizationColor

### PR DESCRIPTION
**Pull request by MarshallOfSound**
This is useful for automatic app theming

* Fetches the color from the registry
* Returns an empty string if the fetch fails (a falsey value)

### CODE REVIEW

File: docs/api/app.md [hunk 1]
```diff
@@ -852,6 +852,19 @@ correctly.
 
 **Note:** This will not affect `process.argv`.
 
+### `app.getColorizationColor()` _Windows_
+
+Returns the users current system wide color preference in the form of an ARGB
```
Comment (paulcbetts) [inline docs/api/app.md]: Maybe include a screenshot of where this color is used?

Comment (MarshallOfSound) [inline docs/api/app.md]: Inside the docs?

Perhaps linking to something like this http://www.windowscentral.com/how-custom-accent-color-windows-10 explains what the color is.

On Windows 7 and 8 it is your Aero theme color.  On Windows 10 it's your "theme color".

Comment (paulcbetts) [inline docs/api/app.md]: Ya, that works too - I'm just not sure that people will know what this is. I'd even go so far as to rename this `getAccentColor`

**Comment by ficristo** [comment]
@MarshallOfSound could you explain more "This is useful for automatic app theming" ?

**Comment by MarshallOfSound** [comment]
@ficristo On Windows 7+ users can specify a theme color.  Some apps offer the ability to either use a custom window frame (using the frameless window) or customize the theme of their app.

It would be nice to utilize the users system color preference as the custom frame or theme hence this method

File: docs/api/app.md [hunk 2]
```diff
@@ -852,6 +852,19 @@ correctly.
 
 **Note:** This will not affect `process.argv`.
 
+### `app.getSystemAccentColor()` _Windows_
+
+Returns the users current system wide color preference in the form of an ARGB
+hexadecimal string.
```
Comment (sindresorhus) [inline docs/api/app.md]: I would return it as RGBA hex to match browsers: http://stackoverflow.com/a/27802062/64949

**Comment by miniak** [comment]
@MarshallOfSound Why don't you use `DwmGetColorizationColor`?

**Comment by MarshallOfSound** [comment]
@miniak Reading from the registry seemed like a better plan than using an undocumented API 👍

**Comment by miniak** [comment]
@MarshallOfSound How is this undocumented? :) https://msdn.microsoft.com/en-us/library/windows/desktop/aa969513(v=vs.85).aspx

**Comment by miniak** [comment]
Btw there should also be a notification triggered by `WM_DWMCOLORIZATIONCOLORCHANGED`

**Comment by MarshallOfSound** [comment]
@miniak Wut, I could have sworn those methods were undocumented 😢 Maybe I'm finally going mad 😆

File: atom/browser/browser.h [hunk 3]
```diff
@@ -167,6 +167,13 @@ class Browser : public WindowListObserver {
   // one from app's name.
   // The returned string managed by Browser, and should not be modified.
   PCWSTR GetAppUserModelID();
+
+  typedef HRESULT (STDAPICALLTYPE *DwmGetColorizationColor)(DWORD *, BOOL *);
```
Comment (miniak) [inline atom/browser/browser.h]: @MarshallOfSound this is unnecessary, Electron only supports Windows 7+

**Comment by MarshallOfSound** [comment]
@miniak Switched to use the DWMAPI dll.  Also added the `system-accent-color-changed` event 👍

File: atom/browser/api/atom_api_app.cc [hunk 4]
```diff
@@ -844,6 +848,8 @@ void App::BuildPrototype(
       .SetMethod("setUserTasks", base::Bind(&Browser::SetUserTasks, browser))
       .SetMethod("getJumpListSettings", &App::GetJumpListSettings)
       .SetMethod("setJumpList", &App::SetJumpList)
+      .SetMethod("getSystemAccentColor",
```
Comment (miniak) [inline atom/browser/api/atom_api_app.cc]: @MarshallOfSound I think this belongs to the new systemPreferences module

File: atom/browser/browser.h (hunk 3 above)
Comment (MarshallOfSound) [inline atom/browser/browser.h]: Compile failed without it so I had to add it in.  I didn't think it was necessary but ninja disagreed 😆

Comment (miniak) [inline atom/browser/browser.h]: What about fixing the build instead of adding complexity to the code? :) what was it complaining about?

Comment (MarshallOfSound) [inline atom/browser/browser.h]: Basically a whole bunch of unresolved identifiers, it didn't know what `DwmGetColorizationColor` was.  I'll try harder to work around it but adding one `typedef` didn't seem to bad at the time 👍

File: atom/browser/api/atom_api_app.cc (hunk 4 above)
Comment (MarshallOfSound) [inline atom/browser/api/atom_api_app.cc]: Moving the method to `systemPreferences` is relatively easy.  The issue is that the event `WM_DWMCOLORIZATIONCOLORCHANGED` is emitted at the window level, and I don't think we should be coupling the `NativeWindow` class with `systemPreferences`.

File: atom/browser/native_window_views_win.cc [hunk 5]
```diff
@@ -90,6 +90,15 @@ bool NativeWindowViews::PreHandleMSG(
     // mode if it isn't already, always say we didn't handle the message
     // because we still want Chromium to handle returning the actual
     // accessibility object.
+    case WM_DWMCOLORIZATIONCOLORCHANGED: {
```
Comment (zcbenz) [inline atom/browser/native_window_views_win.cc]: Putting this in `NativeWindowViews` makes the event get emitted for multiple times when there are multiple windows. The common way is to create [message window](https://msdn.microsoft.com/en-us/library/windows/desktop/ms632599(v=vs.85).aspx#message_only) to receive the messages, and there is a helper class in Chromium to do it: https://cs.chromium.org/chromium/src/base/win/message_window.h?sq=package:chromium.

By using it we can also move the API to `systemPreferences`.

Comment (MarshallOfSound) [inline atom/browser/native_window_views_win.cc]: @zcbenz I'm unfamiliar with that helper class so I'm probable doing something silly.  But while trying to do this

```cpp
CreateNamed(base::Bind(&SystemPreferences::HandleMessage, this), L"test");
```

I get this glorious error 

```bash
vendor\brightray\vendor\download\libchromiumcontent\src\base\memory\ref_counted.h(407): error C2039: 'Release': is not a member of 'atom::api::SystemPreferences'
```

Is there another helper that provides the required `Release` and `AddRef` methods that making a pointer to a non static member function requires?

Comment (zcbenz) [inline atom/browser/native_window_views_win.cc]: Usually you can do this to pass raw pointers:

```c++
base::Bind(&SystemPreferences::HandleMessage, base::Unretained(this))
```

Comment (MarshallOfSound) [inline atom/browser/native_window_views_win.cc]: @zcbenz That works, thanks.

However, after some more exploration it looks like a message only window won't do the job as 

> does not receive broadcast messages

https://msdn.microsoft.com/en-us/library/windows/desktop/ms632599.aspx#message_only

Comment (zcbenz) [inline atom/browser/native_window_views_win.cc]: I'm not sure whether `WM_DWMCOLORIZATIONCOLORCHANGED` belongs to broadcast message, most system notifications do not.

If it is we have to use a hidden window to deal with it.

Comment (MarshallOfSound) [inline atom/browser/native_window_views_win.cc]: @zcbenz Updated to create a hidden window inside the `systemPreferences` module

File: atom/browser/api/atom_api_system_preferences_win.cc [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+#include "base/win/wrapped_window_proc.h"
+#include "ui/base/win/shell.h"
+#include "ui/gfx/win/hwnd_util.h"
+
+namespace atom {
+
+namespace {
+
+const wchar_t kSystemPreferencestWindowClass[] =
```
Comment (levrik) [inline atom/browser/api/atom_api_system_preferences_win.cc]: Is `Preferencest` a typo?

**Comment by zcbenz** [comment]
:+1:

**Comment by paulcbetts** [comment]
Thanks @MarshallOfSound! This is great

**Pull request merged**